endpoints/openrtb2/auction.go

```diff
@@ -218,10 +248,88 @@ func (deps *endpointDeps) validateImpExt(ext openrtb.RawJSON, impIndex int) erro
 			if err := deps.paramsValidator.Validate(bidderName, ext); err != nil {
 				return fmt.Errorf("request.imp[%d].ext.%s failed validation.\n%v", impIndex, bidder, err)
 			}
-		} else {
+		} else if bidder != "prebid" {
 			return fmt.Errorf("request.imp[%d].ext contains unknown bidder: %s", impIndex, bidder)
 		}
 	}
 
 	return nil
 }
+
+// Check the request for configs, patch in the stored config if found.
+func (deps *endpointDeps) processConfigs(ctx context.Context, request *openrtb.BidRequest) []error {
+	// Potentially handle Request level configs.
+
+	// Pull the Imp configs.
+	configIds, shortIds, errList := deps.findImpConfigIds(request.Imp)
+
+	configs, errL := deps.configFetcher.GetConfigs(ctx, shortIds)
+	if len(errL) > 0 {
+		errList = append(errList, errL...)
+	}
+
+	// Process Imp level configs.
+	for i := 0; i < len(request.Imp); i++ {
+		// Check if a config was requested
+		if len(configIds[i]) > 0 {
+			conf, ok := configs[configIds[i]]
+			if ok && len(conf) > 0 {
+				err := deps.processImpConfig(&request.Imp[i], conf)
+				if err != nil {
+					errList = append(errList, err)
+				}
+			}
+		}
+	}
+	if len(errList) == 0 {
+		return nil
+	}
+	return errList
+}
+
+// Pull the Imp configs. Return both ID indexed by Imp array index, and a simple list of existing IDs.
+func (deps *endpointDeps) findImpConfigIds(imps []openrtb.Imp) ([]string, []string, []error) {
+	errList := make([]error, 0, len(imps))
+	configIds := make([]string, len(imps))
+	shortIds := make([]string, 0, len(imps))
+	for i := 0; i < len(imps); i++ {
+		if imps[i].Ext != nil && len(imps[i].Ext) > 0 {
+			eConf := &impExtConfig{}
+			err := json.Unmarshal(imps[i].Ext, eConf)
+			if err == nil && len(eConf.Prebid.Managedconfig.ID) > 0 {
+				configIds[i] = eConf.Prebid.Managedconfig.ID
+				shortIds = append(shortIds, eConf.Prebid.Managedconfig.ID)
+			} else if len(eConf.Prebid.Managedconfig.ID) > 0 {
+				errList = append(errList, err)
+				configIds[i] = ""
+			}
+		} else{
+			configIds[i] = ""
+		}
+	}
+	return configIds, shortIds, errList
+}
+
+
+// Process the configs for an Imp. Need to modify the Imp object in place as we cannot simply assign one Imp
+// to another (deep copy)
+func (deps *endpointDeps) processImpConfig(imp *openrtb.Imp, conf json.RawMessage) error {
+	impJson, err := json.Marshal(imp)
+	if err != nil {
+		return err
+	}
+
+	newImp, err := jsonpatch.MergePatch(conf, impJson)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(newImp, imp)
+	// Due to the definition of openrtb.Imp, ID will always have a value, even if empty. So we need to restore the ID
+	// from the config if overriden by a blank ID.
```

Comment:
I think this actually highlights a slightly subtle bug...

Once JSON has been parsed into a Go object, we can no longer tell whether the imp was `{ id: "" }`, ` { id: null }`, or ` { ... no id defined ... }`.

It looks like this specialized logic fixes the last case properly... but actually breaks the other two. I'm a bit concerned too because `Imp.ID` isn't the _only_ field where this happens. Other required fields inside the `Imp` (e.g. `imp.video.mimes`, `imp.metric.type`) will have the same problem.

The root cause seems to be that the Request JSON is unmarshalled into a Go object, and then re-marshalled into JSON again. If we can find a way to merge configs without marshalling, everything would work properly and this logic wouldn't be necessary.

I'm not sure there's a great way to do this... but I did find a library which might help? It appears to offer very fast JSON-traversals without having to Unmarshal anything: https://github.com/buger/jsonparser